### PR TITLE
Add `attributes_for_database` method to return attributes as they would be in the database

### PR DESCRIPTION
**Pull request by shioyama**
### Summary

_TL;DR: I wanted a method that returns a record's attributes such that they can be used to regenerate the record with `instantiate`. There is no such method AFAICT, so I added one._

I am working on serialization which takes a record's attributes and serializes them in such a way that they can be used to recreate the record. The criteria is that the serializer should behave externally exactly like `Marshal`.

I started using `foo.attributes_before_type_cast` to serialize, and `Foo.instantiate(attributes_before_type_cast)` to deserialize, but found that this broke on json attributes if the assigned key values are symbols, since in this case `foo.attributes_before_type_cast` will return the attributes with symbol keys, whereas `Marshal.load(Marshal.dump(...))` would return the instance with string-valued keys on the json column.

The problem is that `attributes_before_type_cast`, as its name implies, returns the attributes before any type casting, but what we want to pass to `instantiate` is the attributes _as they would be in the database_.

In this PR I have added a method, `attributes_for_database`, which does this. It has this unique "round-trip" property:

```ruby
Foo.instantiate(foo.attributes_for_database).attributes == foo.attributes
```

In other words, you can use this version of attributes to re-create the original record. This makes it ideal for use in serialization.

I also added `ActiveModel::AttributeSet#values_for_database` which seemed like the right place to actually do the "work" of transforming values.

### Other Information

Not sure about placement of methods and tests, happy to move just let me know 🙏 

Note: there is a precedent for this method in #40456, which added a private `attribute_for_database` method for individual attributes.

cc @rafaelfranca @byroot

### CODE REVIEW

File: activerecord/test/cases/enum_test.rb [hunk 1]
```diff
@@ -378,6 +378,14 @@ class EnumTest < ActiveRecord::TestCase
     assert_equal "published", @book.status
   end
 
+  test "attributes_for_database" do
+    assert_equal 2, @book.attributes_for_database["status"]
+
+    @book.status = "published"
+
+    assert_equal 2, @book.attributes_for_database["status"]
+  end
```
Comment (shioyama) [inline activerecord/test/cases/enum_test.rb:387]: This might be overkill, but I added it here since the test above for `attribute_for_database` is very similar and I wanted to follow the existing pattern.

**Comment by casperisfine** [comment]
Make sense to me, I ran into the exact same issue last week with http://github.com/Shopify/identity_cache, so I'm 👍 

I'd love @rafaelfranca's 👀 thought.

**Comment by casperisfine** [comment]
You'll likely need to add this to the changelog.

File: activerecord/lib/active_record/attribute_methods/before_type_cast.rb [hunk 2]
```diff
@@ -66,6 +66,12 @@ def attributes_before_type_cast
         @attributes.values_before_type_cast
       end
 
+      # Returns a hash of attributes for assignment to the database.
+      #
```
Comment (eileencodes) [inline activerecord/lib/active_record/attribute_methods/before_type_cast.rb:70]:
```suggestion
```

**Review by marivaldo — APPROVED** [review]

**Comment by casperisfine** [comment]
@eileencodes are you good with this? If you don't have objections I'll merge it.

**Review by eileencodes — APPROVED** [review]
Yup this looks good to me.
